From bcabadacd10a97429abbb4c8b105983373069c9e Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Thu, 25 Aug 2016 20:50:07 +0800
Subject: [PATCH 68/85] MLK-13119-3 cpufreq: imx6q: support VPU 396MHz

When VPU runs at 396MHz, VDDSOC_CAP's voltage
should be set to 1.275V for all set-points,
add VPU clock rate check to support this case.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
[zou:Original patch taken from
git://git.freescale.com/imx/linux-imx.git imx_4.1.15_2.0.0_ga]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 drivers/cpufreq/imx6q-cpufreq.c |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/drivers/cpufreq/imx6q-cpufreq.c b/drivers/cpufreq/imx6q-cpufreq.c
index 68ff27c..ab19b34 100644
--- a/drivers/cpufreq/imx6q-cpufreq.c
+++ b/drivers/cpufreq/imx6q-cpufreq.c
@@ -273,11 +273,13 @@ static int imx6q_cpufreq_probe(struct platform_device *pdev)
 {
 	struct device_node *np;
 	struct dev_pm_opp *opp;
+	struct clk *vpu_axi_podf;
 	unsigned long min_volt, max_volt;
 	int num, ret;
 	const struct property *prop;
 	const __be32 *val;
 	u32 nr, j, i = 0;
+	u32 vpu_axi_rate = 0;
 
 	cpu_dev = get_cpu_device(0);
 	if (!cpu_dev) {
@@ -312,6 +314,10 @@ static int imx6q_cpufreq_probe(struct platform_device *pdev)
 	secondary_sel = devm_clk_get(cpu_dev, "secondary_sel");
 	osc = devm_clk_get(cpu_dev, "osc");
 
+	vpu_axi_podf = devm_clk_get(cpu_dev, "vpu_axi_podf");
+	if (!IS_ERR(vpu_axi_podf))
+		vpu_axi_rate = clk_get_rate(vpu_axi_podf);
+
 	arm_reg = devm_regulator_get_optional(cpu_dev, "arm");
 	pu_reg = devm_regulator_get_optional(cpu_dev, "pu");
 	soc_reg = devm_regulator_get(cpu_dev, "soc");
@@ -397,6 +403,13 @@ static int imx6q_cpufreq_probe(struct platform_device *pdev)
 					imx6_soc_volt[soc_opp_count - 1] = 1250000;
 				}
 #endif
+				if (vpu_axi_rate == 396000000) {
+					if (freq <= 996000) {
+						pr_info("increase SOC/PU voltage for VPU396MHz at %ld MHz\n",
+							freq / 1000);
+						imx6_soc_volt[soc_opp_count - 1] = 1275000;
+					}
+				}
 				break;
 			}
 		}
-- 
1.7.5.4

